contracts/cw721-stake/src/contract.rs

```diff
@@ -0,0 +1,1821 @@
+    if let Some(reward_token_address) = config.reward_token_address {
+        if info.sender != reward_token_address {
+            return Err(ContractError::InvalidToken {
+                received: info.sender,
+                expected: reward_token_address,
+            });
+        }
+
+        let sender = deps.api.addr_validate(&wrapper.sender)?;
+        REWARD_BALANCE.update(deps.storage, |old| {
+            old.checked_add(wrapper.amount).map_err(StdError::overflow)
+        })?;
+
+        Ok(Response::new()
+            .add_attribute("action", "fund")
+            .add_attribute("from", &sender)
+            .add_attribute("amount", wrapper.amount))
+    } else {
+        Err(ContractError::NothingToFund {})
+    }
+}
+
+pub fn execute_stake(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    wrapper: Cw721ReceiveMsg,
+) -> Result<Response, ContractError> {
+    let config = CONFIG.load(deps.storage)?;
+    if info.sender != config.nft_address {
+        return Err(ContractError::InvalidToken {
+            received: info.sender,
+            expected: config.nft_address,
+        });
+    }
+
+    let sender = deps.api.addr_validate(&wrapper.sender)?;
+    STAKED_NFTS_PER_OWNER.update(
+        deps.storage,
+        &sender,
+        env.block.height,
+        |nft_collection| -> StdResult<HashSet<String>> {
+            let mut updated_nft_collection = nft_collection.unwrap_or_default();
+            updated_nft_collection.insert(wrapper.token_id.clone());
+            Ok(updated_nft_collection)
+        },
+    )?;
+
+    TOTAL_STAKED_NFTS.update(
+        deps.storage,
+        env.block.height,
+        |total_staked_nfts| -> StdResult<HashSet<String>> {
+            let mut updated_total_staked_nfts = total_staked_nfts.unwrap_or_default();
+            updated_total_staked_nfts.insert(wrapper.token_id.clone());
+            Ok(updated_total_staked_nfts)
+        },
+    )?;
+
+    let hook_msgs = stake_hook_msgs(deps.storage, sender.clone(), wrapper.token_id.clone())?;
+    Ok(Response::new()
+        .add_submessages(hook_msgs)
+        .add_attribute("action", "stake")
+        .add_attribute("from", sender)
+        .add_attribute("token_id", wrapper.token_id))
+}
+
+pub fn execute_unstake(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    token_id: String,
+    reward_wallet_address: Option<String>,
+) -> Result<Response, ContractError> {
+    let validated_reward_wallet = reward_wallet_address
+        .map(|h| deps.api.addr_validate(&h))
+        .transpose()?;
+    let sender = info.sender.clone();
+    let config = CONFIG.load(deps.storage)?;
+    let sender_staked_nft_collection = STAKED_NFTS_PER_OWNER
+        .may_load(deps.storage, &sender)?
+        .unwrap_or_default();
+    if !sender_staked_nft_collection.contains(&token_id) {
+        return Err(ContractError::Unauthorized {});
+    }
+    let balance = REWARD_BALANCE.load(deps.storage).unwrap_or_default();
```

Comment:
Because we initialize this to zero in instantiate might be best to error here if `load` fails. This hitting the default case here suggests to me that something has gone wrong.